src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2StreamWindowManager.cs

```diff
@@ -0,0 +1,151 @@
+                : base(connection, stream)
+            {
+                _magic = connection._pool.Settings._streamWindowMagicMultiplier;
+                _stream.Trace($"magic:{_magic}");
+            }
+
+            public override void AdjustWindow(int bytesConsumed)
+            {
+                _delivered += bytesConsumed;
+                if (_delivered < StreamWindowThreshold)
+                {
+                    return;
+                }
+
+                TimeSpan rtt = _connection._rttEstimator!.Rtt;
+                TimeSpan dt = DateTime.Now - _lastWindowUpdate;
+
+                int windowSizeIncrement = _delivered;
+
+                if (_magic * _delivered * rtt.Ticks > StreamWindowThreshold * dt.Ticks)
+                {
+                    windowSizeIncrement += _streamWindowSize;
+                    _streamWindowSize *= 2;
+
+                    _stream.Trace($"Updated StreamWindowSize: {StreamWindowSize}, StreamWindowThreshold: {StreamWindowThreshold}");
+                }
+                else
+                {
+                    string msg =
+                        $"No adjustment! | RTT={rtt.TotalMilliseconds} ms || dt={dt.TotalMilliseconds} ms || " +
+                        //$"_delivered * rtt.Ticks = {_delivered * rtt.Ticks} || StreamWindowThreshold * dt.Ticks = {StreamWindowThreshold * dt.Ticks} ||" +
+                        $"Magic*_delivered/dt = {_magic* _delivered / dt.TotalSeconds} bytes/sec || StreamWindowThreshold/RTT = {StreamWindowThreshold / rtt.TotalSeconds} bytes/sec";
+                    _stream.Trace(msg);
+                }
+
+                Task sendWindowUpdateTask = _connection.SendWindowUpdateAsync(_stream.StreamId, windowSizeIncrement);
+                _connection.LogExceptions(sendWindowUpdateTask);
+
+                _delivered = 0;
+                _lastWindowUpdate = DateTime.Now;
+            }
+        }
+
+        private class RttEstimator
+        {
+            private readonly TimeSpan _initialRtt;
+            private Http2Connection _connection;
+            private readonly Socket? _socket;
+            public TimeSpan Rtt { get; private set; }
+            public RttEstimator(Http2Connection connection, TimeSpan? fakeRtt, Socket? socket)
+            {
+                _connection = connection;
+                if (fakeRtt.HasValue)
+                {
+                    Rtt = fakeRtt.Value;
+                    _initialRtt = Rtt;
+                }
+                _socket = socket;
+                UpdateEstimation();
+            }
+
+#if WINDOWS
+            internal void UpdateEstimation()
+            {
+                if (_socket == null) return;
+
+                if (Interop.Winsock.GetTcpInfoV0(_socket.SafeHandle, out Interop.Winsock._TCP_INFO_v0 tcpInfo) == SocketError.Success)
+                {
+                    Rtt = TimeSpan.FromTicks(10 * tcpInfo.RttUs);
```

Comment:
We get this value from winsock native call. But we compare it with `DateTime.Now` diffs. I know that the unit here is correct, but the different types of time measuring might have some differences, so we might be comparing apples to pears.

But you said you hit the limit on Linux machine, I assume that there is only `fakeRtt` available, so this might not have any significant effect.